Add custom healpix plotting method

smcguire-cmu opened the pull request:
Replaces healpy's plotting methods with a new healpix map plotting method based on mocpy's plotting functions.

Fixes #242

hombit commented on src/hipscat/inspection/visualize_catalog.py:
```diff
@@ -29,118 +47,247 @@ def _read_point_map(catalog_base_dir):
+        )
+    _plot_healpix_value_map(ipix, depth, healpix_map, ax, wcs, cmap=cmap, norm=norm, cbar=cbar, **kwargs)
+    plt.grid()
+    plt.ylabel("dec")
+    plt.xlabel("ra")
```
```suggestion
    plt.ylabel("Dec")
    plt.xlabel("RA")
```
or 
```suggestion
    plt.ylabel(r"$\delta$")
    plt.xlabel(r"$\alpha$")
```

hombit commented on src/hipscat/inspection/visualize_catalog.py:
```diff
@@ -29,118 +47,247 @@ def _read_point_map(catalog_base_dir):
+    return ipix_d
+
+
+def plot_healpix_map(
```
HIgh-level question. If I select some FOV, the color bar still shows the same range as for the whole plot. I'd probably prefer to have better color resolution when I zoom. Maybe it is just a documentation issue about `norm`/`cmap` arguments.

See an example with `plot_points` and DES DR2 catalog:
<img width="843" alt="image" src="https://github.com/user-attachments/assets/e40f5cc1-de4a-44d6-908c-fb64f9af5449">
<img width="801" alt="image" src="https://github.com/user-attachments/assets/e884e604-ccad-47a0-8c72-c8c80c3ea068">

smcguire-cmu replied:
Yeah, we should cull the pixels outside of the fov, but I was lazy and I didn't think about this use case so I didn't think it would impact anything other than performance. But this is a good idea to add.

smcguire-cmu replied:
Ok this has been updated! Pixels outside the FOV now get culled

hombit reviewed (approved):
Great work, Sean!

hombit commented:
@smcguire-cmu I played more! It gave me empty plot when I tried to do it for ALLWISE:

```python
import hipscat

catalog = hipscat.loaders.read_from_hipscat('https://data.lsdb.io/unstable/wise/allwise')
hipscat.inspection.plot_points(catalog)
```
<img width="799" alt="image" src="https://github.com/user-attachments/assets/e449781c-5de8-495c-bb4d-c5d20045ae3f">

It also uses a lot of RAM, but it is probably expected?

Zoomed version works fine

```python
import hipscat
from astropy.coordinates import Angle, SkyCoord

catalog = hipscat.loaders.read_from_hipscat('https://data.lsdb.io/unstable/wise/allwise')
hipscat.inspection.plot_points(catalog, center=SkyCoord(0, 0, unit='deg'), fov=(Angle(5, 'deg'), Angle(10, 'deg')))
```
<img width="800" alt="image" src="https://github.com/user-attachments/assets/0c2d8811-20ca-4572-b078-e2c7a5d38024">

smcguire-cmu commented on src/hipscat/inspection/visualize_catalog.py:
```diff
@@ -29,118 +47,247 @@ def _read_point_map(catalog_base_dir):
     Args:
         pixels: healpix pixels (order and pixel number) to visualize
         plot_title (str): heading for the plot
-        projection (str) The map projection to use. Valid values include:
-            - moll - Molleweide projection (default)
-            - gnom - Gnomonic projection
-            - cart - Cartesian projection
-            - orth - Orthographic projection
```
The string works differently now, but I forgot to add in the docstring for this arg.

smcguire-cmu commented on src/hipscat/inspection/visualize_catalog.py:
```diff
@@ -29,118 +47,247 @@ def _read_point_map(catalog_base_dir):
+            too_large_vals = backfacing_vals
+
+        next_depth = depth + 1
+        ipixels = np.array([], dtype=ipixels.dtype)
+        vals = np.array([], dtype=vals.dtype)
```
Yeah, if there's no next depth then we default to an empty array. I'll add a comment for it

hombit commented:
@smcguire-cmu One more thing. It would be nice to make a "view window" to be a rectangle when `fov` is passed

smcguire-cmu commented on src/hipscat/inspection/visualize_catalog.py:
```diff
@@ -29,118 +47,247 @@ def _read_point_map(catalog_base_dir):
+        if ax is not None:
+            fig = ax.get_figure()
+        else:
+            fig = plt.figure(figsize=(10, 5))
```
I think we talked about this and agreed that it's fine for the user to pass in an axes if they want to overplay an existing plot

smcguire-cmu commented on src/hipscat/inspection/visualize_catalog.py:
```diff
@@ -29,118 +47,247 @@ def _read_point_map(catalog_base_dir):
+    # Set projection
+    _set_wcs(ax, wcs)
+
+    return ax
```
I like `plot_healpix_map` returning ax, it makes it easy to overplot things, and you can get the collection from it anyway. You're right though that the inner function doesn't really need to return anything, I'll update that.

smcguire-cmu replied:
I also like that it matches plt.subplots `fig, ax = plt.subplots()`

smcguire-cmu commented on tests/hipscat/inspection/test_visualize_catalog.py:
```diff
@@ -1,91 +1,477 @@
+    non_edge_pixels = {}
+    pixels_ord3 = np.arange(4**3 * order_0_pix, 4**3 * (order_0_pix + 1))
+    non_edge_pixels[3] = pixels_ord3[~np.isin(pixels_ord3, edge_pixels[3])]
+    for o in range(4, 8):
```
yeaaah, I just hate that ord is a keyword and I always have order defined already. But I'll fix it, thanks

smcguire-cmu commented on src/hipscat/inspection/visualize_catalog.py:
```diff
@@ -29,118 +47,247 @@ def _read_point_map(catalog_base_dir):
+        else:
+            fig = plt.figure(figsize=(10, 5))
+    if fov is None:
+        fov = (320 * u.deg, 160 * u.deg)
```
This is how you specify the FOV for full sky. It's a bit janky that it's not 360, 180, but they say that they know it's not perfect on mocpy.

I guess it is immutable, but it feels a bit big to me to put in the function definition. Same with the skycoord center. That might be me being weird though. I agree that at least putting it in the doctoring is a good idea though.

smcguire-cmu commented:
> @smcguire-cmu I played more! It gave me empty plot when I tried to do it for ALLWISE:
> 
> ```python
> import hipscat
> 
> catalog = hipscat.loaders.read_from_hipscat('https://data.lsdb.io/unstable/wise/allwise')
> hipscat.inspection.plot_points(catalog)
> ```
> 
> <img alt="image" width="799" src="https://private-user-images.githubusercontent.com/1784493/377204838-e449781c-5de8-495c-bb4d-c5d20045ae3f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3MTU0OTAsIm5iZiI6MTcyOTcxNTE5MCwicGF0aCI6Ii8xNzg0NDkzLzM3NzIwNDgzOC1lNDQ5NzgxYy01ZGU4LTQ5NWMtYmI0ZC1jNWQyMDA0NWFlM2YucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjNUMjAyNjMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjFmYmQzZmVmMzM3YTk5Nzc5Y2FlZGIxNjE2OTQ2MDA3ZGI0YTZjYjRhOWIxYzg2NjExODQwNWYxZDU4ZGUyOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.XotS-UpbSdu-0IqGCw4ljzaA6dz7MxAjLOVexF0KinA">
> It also uses a lot of RAM, but it is probably expected?
> 
> Zoomed version works fine
> 
> ```python
> import hipscat
> from astropy.coordinates import Angle, SkyCoord
> 
> catalog = hipscat.loaders.read_from_hipscat('https://data.lsdb.io/unstable/wise/allwise')
> hipscat.inspection.plot_points(catalog, center=SkyCoord(0, 0, unit='deg'), fov=(Angle(5, 'deg'), Angle(10, 'deg')))
> ```
> 
> <img alt="image" width="800" src="https://private-user-images.githubusercontent.com/1784493/377204439-0c2d8811-20ca-4572-b078-e2c7a5d38024.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3MTU0OTAsIm5iZiI6MTcyOTcxNTE5MCwicGF0aCI6Ii8xNzg0NDkzLzM3NzIwNDQzOS0wYzJkODgxMS0yMGNhLTQ1NzItYjA3OC1lMmM3YTVkMzgwMjQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjNUMjAyNjMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2M0NTgyNzM3ZGJiMWQxNjczNWJhNDU2ZTRhNWU5NzY3YzU0M2Y0NmUxNzMyYTYxMTgyN2ZjODU5NTcyNmY2NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.VMVnChTeRGL5q7OC8RDsHlxDQWY0XuA55XhGGX9axc0">

This was an interesting issue! Allwise has order 10 pixels in the point map, which are much smaller than a single pixel in the plot, so the part of the code that culls the backfacing cells detected them as invalid pixels and removed them all. I added some more functionality that merges any healpix pixels smaller than a single plot pixel. There's lots of ways that you can deal with aggregating multiple map pixels, but that's complicated. I figured a rendering engine with multiple objects in the same pixel would just pick one, so for now the behavior is to sample from the map and throw a warning.

hombit reviewed (approved):
This is super-mega great, thank you!
